DAOS-17772 rebuild: refine some rebuild/EC agg/EC deg fetch process

liuxuezhao opened the pull request:
1. add container RDB KV ds_cont_prop_ec_agg_eph for EC aggregation epoch boundary, store it when bump and load it after restart.
2. synchronize the ec agg boundary before rebuild
3. change ec agg boundary IV refresh to be EAGER sync
4. wait discard's completion in ds_pool_tgt_discard_handler()
5. fix a cart IV sync bug that ignored GRP_VER err case's err code
6. fetch ec agg boundary IV in fetch RPC handler when it is zero
7. fix an EC agg partial update stripe processing bug 
8. fix a sgl sg_nr_out and iov_len process bug 
9. fix a EC degraded fetch bug related with punch extent
10. some refines for rebuild upbound epoch
    1) rename rebuild stable epoch to rebuild upbound epoch
    2) select "current hlc + epsilon" (ds_rebuild_get_upbound_eph())
        as the rebuild upbound epoch.
    3) use the global consistent rebuild upbound epoch as rebuild fence
        to avoid VOS aggregation cross the epoch.
   


### Steps for the author:

* [ ] Commit message follows the [guidelines](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Appropriate [Features or Test-tag](https://daosio.atlassian.net/wiki/spaces/DC/pages/10984259629/Test+Tags) pragmas were used.
* [ ] Appropriate [Functional Test Stages](https://daosio.atlassian.net/wiki/spaces/DC/pages/12147556353/CI+Functional+Test+Stages) were run.
* [ ] At least two positive code reviews including at least one code owner from each category referenced in the PR.
* [ ] Testing is complete. If necessary, forced-landing label added and a reason added in a comment.

#### After all prior steps are complete:
* [ ] Gatekeeper requested (daos-gatekeeper added as a reviewer).

liuxuezhao committed:
DAOS-17772 rebuild: refine EC agg epoch boundary process

1. add container RDB KV ds_cont_prop_ec_agg_eph for EC aggregation epoch
   boundary, store it when bump and load it after restart.
2. synchronize the ec agg boundary before rebuild
3. wait discard's completion in ds_pool_tgt_discard_handler()
4. fix a crt IV bug that ignored IV sync's GRP_VER err case

Signed-off-by: Xuezhao Liu <xuezhao.liu@hpe.com>

liuxuezhao commented on src/container/container_iv.c:
```diff
@@ -546,8 +546,38 @@ cont_iv_ent_fetch(struct ds_iv_entry *entry, struct ds_iv_key *key,
+				uint64_t ec_agg_eph;
+
+				rc = ds_cont_ec_agg_eph_rdb_lookup(entry->ns->iv_pool_uuid,
+								   civ_key->cont_uuid, &ec_agg_eph);
```
on rebuild it calls ds_cont_fetch_ec_agg_boundary for each cont when rebuilt the cont and possibly will go to here on PS leader.
One example is on full system restart and rebuild started, all local cache is invalid and the IV fetch ec_agg_boundary will go to PS leader, trigger the code here.

liuxuezhao commented on src/container/srv_container.c:
```diff
@@ -1820,112 +1822,256 @@ ds_cont_tgt_refresh_agg_eph(uuid_t pool_uuid, uuid_t cont_uuid,
+	}
+
+	d_iov_set(&value, &ec_agg_eph, sizeof(ec_agg_eph));
+	rc = rdb_tx_update(&tx, &cont->c_prop, &ds_cont_prop_ec_agg_eph, &value);
```
I thought about it, not very sure if it is good to lookup+compare before update, but before calling cont_agg_eph_store it checked "if (min_eph > ec_agg->ea_rdb_eph)" and only call it when it is true.

liw commented on src/rebuild/srv.c:
```diff
@@ -2180,7 +2186,7 @@ regenerate_task_of_type(struct ds_pool *pool, pool_comp_state_t match_states, ui
 /* Regenerate the rebuild tasks when changing the leader. */
 int
-ds_rebuild_regenerate_task(struct ds_pool *pool, daos_prop_t *prop)
+ds_rebuild_regenerate_task(struct ds_pool *pool, daos_prop_t *prop, uint64_t delay_sec)
```
[Question] Looks like `delay_sec` will always be 0. Is this change still necessary?

liuxuezhao replied:
right, in previous version it passed in a delay for reboot, in this version i removed that change.
Did not remove the function parameter although, not sure if later can be used.

liw commented on src/container/srv_container.c:
```diff
@@ -1912,15 +1917,30 @@ cont_agg_eph_store(struct cont_svc *svc, uuid_t cont_uuid, uint64_t ec_agg_eph)
+		if (rc == 0)
+			rc = rdb_tx_commit(&tx);
+	} else {
+		D_DEBUG(DB_MD, DF_CONT ": bypass rdb update ec_agg_eph "DF_X64", in rdb eph "DF_X64,
+			DP_CONT(svc->cs_pool_uuid, cont_uuid), ec_agg_eph, old_eph);
```
[Question] In this case, `rc` is 0, and `cont_agg_eph_sync` will set `ea_rdb_eph` to the value of `ec_agg_eph` here, which <= `old_eph`, and then use the lower of the two values for refreshing. Is that okay?

liuxuezhao replied:
I'll check and refine it later after do some test for another issue, thx

Nasf-Fan commented on src/pool/srv_target.c:
```diff
@@ -2438,8 +2438,13 @@ cont_discard_cb(daos_handle_t ih, vos_iter_entry_t *entry,
+#if 0
 	if (rc == 0)
 		rc = ds_cont_child_destroy(arg->tgt_discard->pool_uuid, entry->ie_couuid);
+#endif
```
Some confused. Before incremental reintegration being landed, if without discarding the old container, how can we guarantee the old/stale (punched) object/key can be removed after reintegration?

liuxuezhao replied:
all object already discarded in obj_discard_cb(), here just keep the vos container now.

daosbuild3 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16755/5/execution/node/1349/log

Nasf-Fan commented on src/container/container_iv.c:
```diff
@@ -653,8 +683,18 @@ cont_iv_ent_update(struct ds_iv_entry *entry, struct ds_iv_key *key,
+				/* container non-exist possibly due to in reintegrate the container
+				 * discarded ahead. Ignore such err.
+				 */
+				if (rc == -DER_NONEXIST) {
```
You have already changed `cont_discard_cb()` to prevent such case, so the container will not be discarded for reintegration case?

liuxuezhao replied:
ah right, such code is harmless although, for some orphan container cases may be useful. The comment should change, will change it later if need to refresh for some other issue. Mohamad's test still with a few problems.

kccain commented on src/rebuild/rpc.h:
```diff
@@ -56,7 +57,6 @@ extern struct crt_proto_format rebuild_proto_fmt;
 	((uint32_t)		(rsi_layout_ver)	CRT_VAR)
 
 #define DAOS_OSEQ_REBUILD_SCAN	/* output fields */		 \
-	((uint64_t)		(rso_stable_epoch)	CRT_VAR) \
```
Since the rebuild RPC protocol version (DAOS_REBUILD_VERSION 4) is not being updated perhaps it is safest to keep the uint64_t member in this definition, but rename it to rso_reserved or something like that?

liuxuezhao replied:
as server rolling upgrade does not supported now, the DAOS_REBUILD_VERSION is not really make sense now. 
But I'll add a u64 reservation if need to refresh the PR later.

kccain commented on src/rebuild/rebuild_iv.c:
```diff
@@ -201,12 +201,10 @@ rebuild_iv_ent_refresh(struct ds_iv_entry *entry, struct ds_iv_key *key,
-			       dst_iv->riv_stable_epoch);
+		D_ASSERT(rpt->rt_upbound_eph != 0);
+		if (rpt->rt_upbound_eph != dst_iv->riv_stable_epoch)
+			D_WARN("leader change stable epoch from " DF_U64 " to " DF_U64 "\n",
```
also, will this branch ever be taken now that rpt->rt_stable_epoch does not get updated anymore with the result of the scan broadcast RPC reply rso_stable_epoch ?

liuxuezhao replied:
rso_stable_epoch removed.

kccain replied:
Understood. But does this branch ever get taken?

liuxuezhao commented on src/container/srv_layout.h:
```diff
@@ -137,6 +138,7 @@ extern d_iov_t ds_cont_prop_co_md_times;	/* co_md_times */
 extern d_iov_t ds_cont_prop_cont_obj_version;	/* uint32_t */
 extern d_iov_t ds_cont_prop_nhandles;		/* uint32_t */
 extern d_iov_t ds_cont_prop_oit_oids;		/* snapshot OIT OID KVS */
+extern d_iov_t ds_cont_prop_ec_agg_eph;         /* uint64_t */
```
yes, no problem, as it does not affect any compatibility

daosbuild3 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-16755/16/execution/node/1519/log

liuxuezhao committed:
DAOS-17772 object: fix sgl set bug - sg_nr_out and iov_len

Signed-off-by: Xuezhao Liu <xuezhao.liu@hpe.com>

liuxuezhao committed:
DAOS-17979 object: fix an bug in agg_update_parity

Signed-off-by: Wang Shilong <shilong.wang@hpe.com>

Nasf-Fan commented on src/rebuild/srv.c:
```diff
@@ -1613,6 +1603,13 @@ rebuild_task_ult(void *arg)
+	if (rc) {
+		D_ERROR(DF_UUID ": ds_cont_svc_refresh_agg_eph failed, " DF_RC "\n",
+			DP_UUID(task->dst_pool_uuid), DP_RC(rc));
+		goto out_task;
```
`goto out_pool;`

Nasf-Fan commented on src/rebuild/rebuild_iv.c:
```diff
@@ -201,12 +201,10 @@ rebuild_iv_ent_refresh(struct ds_iv_entry *entry, struct ds_iv_key *key,
+		D_ASSERT(rpt->rt_upbound_eph != 0);
+		if (rpt->rt_upbound_eph != dst_iv->riv_stable_epoch)
+			D_WARN("leader change stable epoch from " DF_U64 " to " DF_U64 "\n",
+			       rpt->rt_upbound_eph, dst_iv->riv_stable_epoch);
```
Confused, which one is changed? The `rt_upbound_eph` is not related with `riv_stable_epoch`, right?

Nasf-Fan commented on src/pool/srv_pool.c:
```diff
@@ -7339,7 +7342,7 @@ pool_svc_update_map(struct pool_svc *svc, crt_opcode_t opc, bool exclude_rank,
 	int				rc;
 	char				*env;
-	daos_epoch_t			rebuild_eph = d_hlc_get();
+	daos_epoch_t                     rebuild_eph = ds_rebuild_get_upbound_eph();
```
Strictly speaking, it is still not safe to generate`rebuild_eph` via adding `epsilon` against current HLC before really refreshing the pool map. Because we do not know how long the refreshing pool map will take. If it takes more time than `epsilon`, then some new update may happen with newer epoch (than `rebuild_eph`) before refreshing pool map completed. And then current reintegration may not process those `new` data. That is unexpected.

On the other hand, since we have `epsilon` windows for reintegration, we need to double check VOS layer overwritten related logic to guarantee that if some data is `prepared` but not real `committed` yet, then the overwritten (from reintegration) with same epoch can be properly processed instead of returning failure.

liuxuezhao replied:
we may need some offline discuss.

Nasf-Fan commented on src/object/srv_obj.c:
```diff
@@ -1483,6 +1502,32 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc, daos_iod_t *io
+
+			arg.eab_pool = ioc->ioc_coc->sc_pool->spc_pool;
+			uuid_copy(arg.eab_co_uuid, ioc->ioc_coc->sc_uuid);
+			rc = dss_ult_execute(obj_fetch_ec_agg_boundary, &arg, NULL, NULL,
```
What is the benefit of creating another ULT to fetch EC agg boundary instead of directly call `obj_fetch_ec_agg_boundary`()?

liuxuezhao replied:
obj_fetch_ec_agg_boundary() can only be executed on system XS 0, but here it is IO service XS.

Nasf-Fan commented on src/object/srv_ec_aggregate.c:
```diff
@@ -1074,7 +1074,7 @@ agg_update_parity(struct ec_agg_entry *entry, uint8_t *bit_map,
 	diff = entry->ae_sgl.sg_iovs[AGG_IOV_DIFF].iov_buf;
 
-	for (i = 0, j = 0; i < cell_cnt; i++) {
+	for (i = 0, j = 0; i < cell_cnt; i++, j++) {
```
How to guarantee the subsequent `while (!isset(bit_map, j)) j++;` will not out of boundary?

Nasf-Fan commented on src/object/cli_ec.c:
```diff
@@ -2677,6 +2679,11 @@ obj_ec_sgl_copy(d_sg_list_t *sgl, uint64_t off, void *buf, uint64_t size)
+	if (sgl_idx.iov_offset == 0)
+		sgl->sg_nr_out = sgl_idx.iov_idx;
+	else
+		sgl->sg_nr_out = sgl_idx.iov_idx + 1;
```
Would you please to comment for the reason? Thanks

Nasf-Fan commented on src/container/srv_container.c:
```diff
@@ -1820,116 +1822,274 @@ ds_cont_tgt_refresh_agg_eph(uuid_t pool_uuid, uuid_t cont_uuid,
-					continue;
-				}
+		/* for reboot case the ea_rdb_eph possibly higher than min_eph */
+		if (min_eph < ec_agg->ea_rdb_eph)
```
Why reboot may make `min_eph` to be older? Because the subsequent `cont_agg_eph_store` may be failed but still go ahead?

liuxuezhao committed:
DAOS-17772 object: fix a data corruption in EC degraded fetch

Differentiate punch record with normal hole for degraded fetch.

Signed-off-by: Xuezhao Liu <xuezhao.liu@hpe.com>

liuxuezhao commented:
@reviewers, sorry that as with too many original commits, I recombined some commits together to make 5 a little bit clearer PRs/commits so did not keep original commit history.

liuxuezhao committed:
DAOS-17772 rebuild: refine rebuild upbound epoch handling

1. rename rebuild stable epoch to rebuild upbound epoch
2. select "current hlc + epsilon" (ds_rebuild_get_upbound_eph())
   as the rebuild upbound epoch.
3. use the global consistent rebuild upbound epoch as rebuild fence
   to avoid VOS aggregation cross the epoch.
4. several naming refine and cleanup.
5. delay rebuild 200S for restart

Signed-off-by: Xuezhao Liu <xuezhao.liu@hpe.com>
